Core/Modes/Matchmode.cs

```diff
@@ -1,43 +1,132 @@
+                }
+                catch (OperationCanceledException)
+                {
+                    _logger.LogInformation("Match loop cancelled");
+                }
             }
+            await _broadcastServer.Stop();
+            await overlayWebsocketTask;
             _logger.LogInformation("Matchmode ended");
         }
 
+        private async Task Loop(CancellationToken cancellationToken)
+        {
+            var teams = new Teams
+            {
+                Blue = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+                Red = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+            };
+            await Task.Delay(TimeSpan.FromSeconds(3), cancellationToken);
+
+            IMatchCycle match = new CoinflipMatchCycle(_loggerFactory.CreateLogger<CoinflipMatchCycle>());
+            Task setupTask = match.SetUp(new MatchInfo(teams.Blue, teams.Red), cancellationToken);
+            await _overlayConnection.Send(new MatchCreatedEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchBettingEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchModesChosenEvent(), cancellationToken); // TODO
+            await _overlayConnection.Send(new MatchSettingUpEvent
+            {
+                MatchId = 1234,
+                Teams = teams,
+                BettingDuration = _matchmodeConfig.DefaultBettingDuration.TotalSeconds,
+                RevealDuration = 0,
+                Gimmick = "speed",
+                Switching = "never",
+                BattleStyle = "singles",
+                InputOptions = new InputOptions
+                {
+                    Moves = new MovesInputOptions
+                    {
+                        Policy = "always",
+                        Permitted = ImmutableList.Create("a", "b", "c", "d")
+                    },
+                    Switches = new SwitchesInputOptions
+                    {
+                        Policy = "never",
+                        Permitted = ImmutableList<string>.Empty,
+                        RandomChance = 0
+                    },
+                    Targets = new TargetsInputOptions
+                    {
+                        Policy = "disabled",
+                        Permitted = ImmutableList<string>.Empty,
+                        AllyHitChance = 0
+                    },
+                },
+                BetBonus = 35,
+                BetBonusType = "bet",
+            }, cancellationToken);
+
+            Duration bettingBeforeWarning = _matchmodeConfig.DefaultBettingDuration - _matchmodeConfig.WarningDuration;
+            await Task.Delay(bettingBeforeWarning.ToTimeSpan(), cancellationToken);
+            await _overlayConnection.Send(new MatchWarningEvent(), cancellationToken);
+
+            await Task.Delay(_matchmodeConfig.WarningDuration.ToTimeSpan(), cancellationToken);
+            await setupTask;
```

Comment:
should be before here, way before

Reply:
i.e. during before warning

Reply:
I don't think so. The setup is meant to be able to run concurrently to the entire betting time. It is awaited as late as possible, which is just before the match is supposed to begin.